src/meta-srv/src/handler.rs

```diff
@@ -103,13 +114,219 @@ impl HeartbeatHandlerGroup {
         let mut acc = HeartbeatAccumulator::default();
         let handlers = self.handlers.read().await;
         for h in handlers.iter() {
-            h.handle(&req, &mut ctx, &mut acc).await?;
+            if ctx.is_skip_all() {
+                break;
+            }
+
+            let role = req.header.as_ref().and_then(|h| Role::from_i32(h.role));
+            if h.is_acceptable(role) {
+                h.handle(&req, &mut ctx, &mut acc).await?;
+            }
         }
         let header = std::mem::take(&mut acc.header);
         let res = HeartbeatResponse {
             header,
-            payload: acc.into_payload(),
+            mailbox_messages: acc.into_mailbox_messages(),
         };
         Ok(res)
     }
 }
+
+struct Tx(oneshot::Sender<Result<MailboxMessage>>);
+struct Rx(oneshot::Receiver<Result<MailboxMessage>>);
+
+pub struct HeartbeatMailbox {
+    pushers: Arc<RwLock<BTreeMap<String, Pusher>>>,
+    sequence: Sequence,
+    senders: DashMap<MessageId, Tx>,
+    recervers: DashMap<MessageId, Rx>,
+    timeouts: DashMap<MessageId, Duration>,
+}
+
+impl HeartbeatMailbox {
+    pub fn with_timeout(
+        pushers: Arc<RwLock<BTreeMap<String, Pusher>>>,
+        sequence: Sequence,
+        check_interval_millis: u64,
+    ) -> MailboxRef {
+        let ins = Arc::new(Self {
+            pushers,
+            sequence,
+            senders: DashMap::default(),
+            recervers: DashMap::default(),
+            timeouts: DashMap::default(),
+        });
+
+        if check_interval_millis > 0 {
+            let timeout_checker = ins.clone();
+            common_runtime::spawn_bg(async move {
```

Comment:
I think we should provide some default value to `check_interval_millis`, to always start the `timeout_checker`. Otherwise the `timeouts` map only gets bigger and bigger.